src/tuner/nccl_ofi_model.c

```diff
@@ -74,180 +66,261 @@ static const float nccl_base_lat[NCCL_NUM_ALGORITHMS][NCCL_NUM_PROTOCOLS] = {
 
-	const double logranks = log(dims->num_ranks) / log(2);
-	const double logbytes = log(nBytes) / log(2);
-	const double lograt = (logbytes - logranks);
-	if (floor(lograt) < 23) {
-		/* Hack until we understand this better. */
-		return INFINITY;
-	}
+	const double internodeLatency =
+		internodeSteps *
+		(base_latency + NCCL_OFI_TUNER_NETWORK_LATENCY +
+		 ((proto == NCCL_PROTO_SIMPLE) ? NCCL_OFI_TUNER_COMPLETION_OVERHEAD : 0));
+	const double intranodeLatency = (base_latency + p2p_latency) * intranodeSteps;
+	const double latencyCost = intranodeLatency + internodeLatency;
 
-	const double steps_per_rotation = dims->num_ranks;
-	const double num_rotations = 2;
-	const double num_steps = (num_rotations * steps_per_rotation) - 1;
-	const double num_steps_internode = (num_rotations * dims->num_nodes);
-	const double num_steps_intranode = num_steps - num_steps_internode;
-
-	double per_proto_bandwidth_overhead;
-	double net_latency = NCCL_OFI_TUNER_NETWORK_LATENCY;
-	switch (proto) {
-	case NCCL_PROTO_LL: {
-		per_proto_bandwidth_overhead = NCCL_OFI_TUNER_LL_PROTO_BW_FACTOR;
-		break;
-	}
-	case NCCL_PROTO_LL128: {
-		per_proto_bandwidth_overhead = NCCL_OFI_TUNER_LL128_PROTO_BW_FACTOR;
-		break;
+	const double internodeBw = 1.0 * NCCL_OFI_TUNER_INTERNODE_BW * chan / 2 / dims->num_ranks;
+	const double intranodeBw = 1.0 * NCCL_OFI_TUNER_INTRANODE_BW * dims->num_ranks;
 
+	double effective_message_size = nBytes;
+	if (proto == NCCL_PROTO_LL) {
+		effective_message_size *= 2;
+	} else if (proto == NCCL_PROTO_LL) {
+		effective_message_size += (nBytes / 15);
 	}
-	case NCCL_PROTO_SIMPLE: {
-		per_proto_bandwidth_overhead = NCCL_OFI_TUNER_SIMPLE_PROTO_BW_FACTOR;
-		net_latency += NCCL_OFI_TUNER_COMPLETION_OVERHEAD;
-		break;
-	}
-	default: {
-		return INFINITY;
-	}
-	}
-
-
-	const double intranode_latency = p2p_latency * num_steps_intranode;
-	const double intranode_bw = (NCCL_OFI_TUNER_INTRANODE_BW);
-	const double intranode_bytes = nBytes * (num_steps_intranode / num_steps);
-	const double intranode_cost = intranode_latency + (intranode_bytes / intranode_bw);
-
-	const double internode_latency = net_latency * num_steps_internode;
-	const double internode_bw = (NCCL_OFI_TUNER_INTERNODE_BW) * per_proto_bandwidth_overhead;
-	const double internode_bytes = nBytes * (num_steps_internode / num_steps);
-	const double internode_cost = internode_latency + (internode_bytes / internode_bw);
 
-	/* XXX: Should be scaled further here, not done for the sake of ensuring
-	 * this is cheaper than NVLSTree at dims where we did not explicitly opt
-	 * out of choosing ring above */
-	return ((base_latency * num_steps) + intranode_cost + internode_cost);
+	double totalCost =
+		((2.0 * effective_message_size / dims->num_ranks) * internodeSteps) / internodeBw;
+	totalCost += ((2.0 * effective_message_size) * intranodeSteps) / intranodeBw;
+	totalCost += latencyCost * (get_ring_num_windows(2 * effective_message_size,
```

Comment:
If we want to add extra latency cost so that that TREE is preferred for small message size, we should do `x2` against the return value of `get_ring_num_windows` instead of `x2` to argument of `get_ring_num_windows`.

Reply:
The cost function compared to real runs of all of these were wildly out of spec, and I'm resolving this conversation to put attention on the adjusted ones.